api/core/v1alpha1/common_helpers.go

```diff
@@ -190,3 +199,46 @@ func findMainContainer(containers []corev1.Container) *corev1.Container {
 	}
 	return nil
 }
+
+// DefaultArgs contain default service args for logservice/dn/tp, these default args set in matrixone-operator-cm configmap
+type DefaultArgs struct {
+	LogService []string `json:"logService,omitempty"`
+	DN         []string `json:"dn,omitempty"`
+	CN         []string `json:"cn,omitempty"`
+}
+
+// setDefaultServiceArgs set default args for service, we only set default args when there is service args config in service spec
+func setDefaultServiceArgs(object interface{}) {
+	if ServiceDefaultArgs == nil {
+		return
+	}
+	switch obj := object.(type) {
+	case *LogSetBasic:
+		// set default arguments only when user does not set any arguments
+		if len(obj.ServiceArgs) == 0 {
+			obj.ServiceArgs = trimEmptyArgs(ServiceDefaultArgs.LogService)
+		}
+	case *DNSetBasic:
+		if len(obj.ServiceArgs) == 0 {
+			obj.ServiceArgs = trimEmptyArgs(ServiceDefaultArgs.DN)
+		}
+	case *CNSetBasic:
+		if len(obj.ServiceArgs) == 0 {
+			obj.ServiceArgs = trimEmptyArgs(ServiceDefaultArgs.CN)
+		}
+	default:
+		moLog.Error(fmt.Errorf("unknown type:%T", object), "expected types: *LogSetBasic, *DNSetBasic, *CNSetBasic")
+		return
+	}
+}
+
+// trimEmptyArgs delete empty string argument "" from slice
+func trimEmptyArgs(args []string) []string {
```

Comment:
As we discussed offline, we don't need to trim empty args, just keep the bad config as is and error the user.